unstructured_ingest/v2/processes/connectors/mongodb.py

```diff
@@ -53,6 +62,166 @@ class MongoDBUploadStagerConfig(UploadStagerConfig):
+class MongoDBIndexer(Indexer):
+    connection_config: MongoDBConnectionConfig
+    index_config: MongoDBIndexerConfig
+    connector_type: str = CONNECTOR_TYPE
+
+    def precheck(self) -> None:
+        """Validates the connection to the MongoDB server."""
+        try:
+            client = self.create_client()
+            client.admin.command("ping")
+        except Exception as e:
+            logger.error(f"Failed to validate connection: {e}", exc_info=True)
+            raise SourceConnectionError(f"Failed to validate connection: {e}")
+
+    @requires_dependencies(["pymongo"], extras="mongodb")
+    def create_client(self) -> "MongoClient":
+        from pymongo import MongoClient
+        from pymongo.driver_info import DriverInfo
+        from pymongo.server_api import ServerApi
+
+        access_config = self.connection_config.access_config.get_secret_value()
+
+        if access_config.uri:
+            return MongoClient(
+                access_config.uri,
+                server_api=ServerApi(version=SERVER_API_VERSION),
+                driver=DriverInfo(name="unstructured", version=unstructured_version),
+            )
+        else:
+            return MongoClient(
+                host=self.connection_config.host,
+                port=self.connection_config.port,
+                server_api=ServerApi(version=SERVER_API_VERSION),
+            )
+
+    @requires_dependencies(["pymongo"], extras="mongodb")
+    def run(self, **kwargs: Any) -> Generator[FileData, None, None]:
+        from bson.objectid import ObjectId
+
+        """Generates FileData objects for each document in the MongoDB collection."""
+        client = self.create_client()
+        database = client[self.connection_config.database]
+        collection = database[self.connection_config.collection]
+
+        # Get list of document IDs
+        ids = collection.distinct("_id")
+        batch_size = self.index_config.batch_size if self.index_config else 100
+        id_batches = [ids[i : i + batch_size] for i in range(0, len(ids), batch_size)]
+
+        for id_batch in id_batches:
+            for doc_id in id_batch:
+                # Handle ObjectId to get the creation time
+                date_created = None
+                if isinstance(doc_id, ObjectId):
+                    date_created = doc_id.generation_time.isoformat()
+
+                source_identifiers = SourceIdentifiers(
+                    fullpath=str(doc_id), filename=str(doc_id), rel_path=str(doc_id)
+                )
+
+                metadata = FileDataSourceMetadata(
+                    date_created=date_created,
+                    date_processed=str(time()),
+                    record_locator={
+                        "database": self.connection_config.database,
+                        "collection": self.connection_config.collection,
+                        "document_id": str(doc_id),
+                    },
+                )
+
+                file_data = FileData(
```

Comment:
Rather than have a single `FileData` for each batch, you could have a single `FileData` of type batch where you store a list of ids. Then in the downloader, you optimize reading the data in batches and expand the first batch `FileData` into single ones. I believe elasticsearch is doing something similar, for an example.

Reply:
Reworked the logic to create one FileData with Ids per batch, then unwrap it in downloader. Same as the elasticsearch does. Good quality catch!